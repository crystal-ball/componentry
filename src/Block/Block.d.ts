import * as React from 'react'
import { SpaceProps } from '../utils/componentry-props'

interface BlockProps {
  inline?: boolean
}

/** https://crystal-ball.github.io/componentry/components/block */
declare const Block: React.ComponentType<BlockProps & SpaceProps>

export default Block
