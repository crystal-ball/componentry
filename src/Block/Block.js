import { useTheme } from '../Theme/Theme'
import { element } from '../factories/element'

/**
 * [Block component 📝](https://componentry.design/components/block)
 */
export function Block(props) {
  const { inline = false, ...rest } = { ...useTheme('Block'), ...props }

  return element({
    componentCx: {
      // ℹ️ Block is a helper component so it doesn't have a 🅲-block class
      'd-block': !inline,
      'd-inline-block': inline,
    },
    ...rest,
  })
}
Block.displayName = 'Block'
