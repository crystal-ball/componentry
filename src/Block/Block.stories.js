/* global module */

import React from 'react'
import { storiesOf } from '@storybook/react'
import { boolean } from '@storybook/addon-knobs'

import Block from './Block'

storiesOf('Block', module).add('<Block />', () => (
  <div>
    <Block inline={boolean('Inline', false)}>Block content</Block>
    <Block inline={boolean('Inline', false)}>Block content</Block>
  </div>
))
