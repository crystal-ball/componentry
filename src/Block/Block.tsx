import React from 'react'
import { useTheme } from '../Theme/Theme'
import { element } from '../factories/element'
import { UtilityProps } from '../utiility-types'

interface BlockProps extends UtilityProps {
  /** Switches between display between an inline and block element */
  inline?: boolean
}

/**
 * [Block component üìù](https://componentry.design/components/block)
 */
export const Block: React.FC<BlockProps> = (props) => {
  const { inline = false, ...rest } = { ...useTheme('Block'), ...props } as BlockProps

  return element({
    componentCx: {
      // ‚ÑπÔ∏è Block is a helper component so it doesn't have a üÖ≤-block class
      'd-block': !inline,
      'd-inline-block': inline,
    },
    ...rest,
  })
}
Block.displayName = 'Block'
