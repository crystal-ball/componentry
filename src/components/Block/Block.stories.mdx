import { ArgsTable, Meta, Story } from '@storybook/addon-docs'

import { Block } from './Block'

export function padNum(num) {
  return num > 9 ? num : `0${num}`
}

<Meta title='Components/Block' component={Block} />

# Block

<Story name='Block'>
  <Block backgroundColor='primary-200' className='rounded'>
    {[1, 2, 3].map((key) => (
      <div
        key={key}
        className='bg-primary-600 text-inverse font-mono font-bold px-6 py-4 rounded'
      >
        #{padNum(key)}
      </div>
    ))}
  </Block>
</Story>

Block is a layout component that renders a div.

<ArgsTable of={Block} />
