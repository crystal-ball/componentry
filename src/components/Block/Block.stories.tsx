import type { Meta, StoryObj } from '@storybook/react'

import { Block } from './Block'

const meta: Meta<typeof Block> = {
  component: Block,
}

export default meta
type Story = StoryObj<typeof Block>

export const Primary: Story = {
  args: {
    backgroundColor: 'primary-200',
    className: 'rounded',
    children: [1, 2, 3].map((num) => (
      <div
        key={num}
        className='bg-primary-600 text-inverse font-mono font-bold px-6 py-4 rounded'
      >
        #{num > 9 ? num : `0${num}`}
      </div>
    )),
  },
}
