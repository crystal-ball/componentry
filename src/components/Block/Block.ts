import { forwardRef } from 'react'
import { type ComponentBaseProps } from '../../utils/base-types'
import { element } from '../../utils/element-creator'
import { useThemeProps } from '../Provider/Provider'

export interface BlockProps extends ComponentBaseProps<'div'> {}

// ‚ú® Nice display type for IntelliSense
export interface Block {
  (props: BlockProps & { ref?: React.ForwardedRef<unknown> }): React.ReactElement | null
  displayName?: string
}

/**
 * #### [üìù Block](https://componentry.design/docs/components/block)
 *
 * Block provides block level layout elements with access to theme utility
 * props.
 * @example
 * ```tsx
 * <Block mt={2} mx={3}>
 *   ...
 * </Block>
 * ```
 */
export const Block: Block = forwardRef((props, ref) => {
  return element({
    ref,
    ...useThemeProps('Block'),
    ...props,
  })
})
Block.displayName = 'Block'
