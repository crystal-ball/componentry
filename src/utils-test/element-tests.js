/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (TestComponent, enzymeOptions = {}) => {
  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', () => {
    // Create a component to validate that the TestComponent returns.
    const TestAs = () => <section />
    const wrapper = mount(<TestComponent as={TestAs} />, enzymeOptions)
    expect(wrapper.find(TestAs).length).toBeTruthy()
  })

  /*
   * All components should spread passed props on the parent container for max
   * customization
   */
  test('should pass through classnames and attributes', () => {
    const wrapper = mount(
      <TestComponent className="test-custom" data-test="test-custom" />,
      enzymeOptions,
    )
    expect(wrapper.find('.test-custom').length).toBeTruthy()
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy()
  })

  // TODO: should render passed children
}
