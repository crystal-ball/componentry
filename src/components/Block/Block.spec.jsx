import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'

import { elementTests } from '../../test/element-tests'
import { Block } from './Block'

describe('<Block />', () => {
  elementTests(Block)
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Block /> snapshots', () => {
  it('renders correctly', () => {
    render(<Block data-testid='block'>Block content</Block>)

    expect(screen.getByTestId('block')).toMatchSnapshot()
  })
})
