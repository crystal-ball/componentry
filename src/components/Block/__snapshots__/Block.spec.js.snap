// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Block /> snapshots renders correctly 1`] = `
<div
  class="block"
  data-testid="block"
>
  Block content
</div>
`;
